File: McbEdu.Mentorias.ShopDemo.Domain.Entities/Customers/Customer.cs
namespace McbEdu.Mentorias.ShopDemo.Domain.Entities.Customers;

public class Customer
{
    public string Name { get; private set; }
    public string Surname { get; private set; }
    public string Email { get; private set; }
    public DateTime BirthDate { get; private set; }

    public Customer(string name, string surname, string email, DateTime birthDate)
    {
        Name = name;
        Surname = surname;
        Email = email;
        BirthDate = birthDate;
    }

    private void ChangeName(string name)
    {
        Name = name;
    }

    private void ChangeSurname(string surname)
    {
        Surname = surname;
    }

    private void ChangeEmail(string email)
    {
        Email = email;
    }

    private void ChangeBirthDate(DateTime birthDate)
    {
        BirthDate = birthDate;
    }
}

File: src/main/java/com/mcbmentorias/shopdemo/domain/entities/customer/Customer.java
package com.mcbmentorias.shopdemo.domain.entities.customer;

import com.mcbmentorias.shopdemo.core.interfaces.IAggregateRoot;
import com.mcbmentorias.shopdemo.domain.entities.base.BaseEntity;
import com.mcbmentorias.shopdemo.domain.entities.customer.inputs.ImportNewCustomerInput;
import com.mcbmentorias.shopdemo.domain.entities.customer.validations.ImportNewCustomerInputValidation;
import lombok.Getter;
import lombok.NoArgsConstructor;

import javax.persistence.Entity;
import javax.persistence.Transient;
import java.time.LocalDate;
import java.util.Objects;
import java.util.UUID;

@Entity
@Getter
@NoArgsConstructor
public class Customer extends BaseEntity implements IAggregateRoot {

    private String name;

    private String lastName;

    private LocalDate birthDate;

    private String email;

    @Transient
    private ImportNewCustomerInputValidation importNewCustomerInputValidation;

    public Customer(
            final ImportNewCustomerInputValidation importNewCustomerInputValidation
    ) {
        this.importNewCustomerInputValidation = importNewCustomerInputValidation;
    }

    public void importCustomer(
        final ImportNewCustomerInput input
    ) {
        if(!this.validate(() -> this.importNewCustomerInputValidation.validate(input))) {
            return;
        }

        this.createNewEntity("import");

        this.setName(input.getName())
            .setLastName(input.getLastName())
            .setBirthDate(input.getBirthDate())
            .setEmail(input.getEmail());
    }

    //Private Method
    private Customer setEmail(final String email) {
        this.email = email;
        return this;
    }

    private Customer setBirthDate(final LocalDate birthDate) {
        this.birthDate = birthDate;
        return this;
    }

    private Customer setLastName(final String lastName) {
        this.lastName = lastName;
        return this;
    }

    private Customer setName(final String name) {
        this.name = name;
        return this;
    }

    public void fill(
            final UUID id,
            final String name,
            final String lastName,
            final LocalDate birthDate,
            final String email
    ) {
        this.fillBase(id);

        this.name = name;
        this.lastName = lastName;
        this.birthDate = birthDate;
        this.email = email;
    }



}